generator/src/test/java/com/scottlogic/deg/generator/restrictions/SetRestrictionsTests.java

```diff
@@ -1,122 +1,199 @@
 package com.scottlogic.deg.generator.restrictions;
 
 import org.junit.Assert;
-import org.junit.jupiter.api.Assertions;
-import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Set;
+import java.util.function.BiConsumer;
 
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.nullValue;
 
 class SetRestrictionsTests {
-    @Test
-    void whenItemIsOnABlacklistAndAWhitelist() {
-        givenFirstInputWhitelist(1, 2, 3);
-        givenSecondInputBlacklist(1);
+    @Nested
+    class fromWhitelist {
```

Comment:
This nested class unit testing is something that I've not seen before. What does it achieve that makes it better than just naming the test methods differently? What does the test report look like? Normally in cases like this I'd name the test method fromWhiteList_withValidSet_usesWhitelistAndEmptyBlacklist

Reply:
Depends which report. In IntelliJ it nests them together nicely in the tree. In other contexts it might look like a failure in:

`com.scottlogic.deg.generator.restrictions.SetRestrictionsTests$fromWhitelist.copiesWhitelistAndUsesEmptyBlacklist`

The main benefits that spring to mind are:

* Nice IDE output
* Less oppressive/repetitious names (`fromWhiteList_withValidSet_usesWhitelistAndEmptyBlacklist` vs `copiesWhitelistAndUsesEmptyBlacklist`)
* Isolation - stateful tests don't bump into each other

The danger is where the class name is separated too much from the test, and you lose so much context that you can't figure out what the test is doing. That's where you might want to put a bit more in the test name, or consider splitting out into a separate test class.